language/vm/vm_runtime/vm_runtime_types/src/value.rs

```diff
@@ -333,93 +367,123 @@ impl Value {
         match value.0 {
-            ValueImpl::Struct(s) => Some(s),
-            _ => None,
+            ValueImpl::Struct(s) => Ok(s),
+            _ => {
+                let msg = format!("Cannot cast {:?} to Struct", value);
+                Err(VMStatus::new(StatusCode::INTERNAL_TYPE_ERROR).with_message(msg))
+            }
         }
     }
 }
 
-impl From<Value> for Option<NativeStructValue> {
-    fn from(value: Value) -> Option<NativeStructValue> {
+impl From<Value> for VMResult<NativeStructValue> {
+    fn from(value: Value) -> VMResult<NativeStructValue> {
         match value.0 {
-            ValueImpl::NativeStruct(s) => Some(s),
-            _ => None,
+            ValueImpl::NativeStruct(s) => Ok(s),
+            _ => {
+                let msg = format!("Cannot cast {:?} to NativeStructValue", value);
+                Err(VMStatus::new(StatusCode::INTERNAL_TYPE_ERROR).with_message(msg))
+            }
         }
     }
 }
 
-impl From<Value> for Option<ReferenceValue> {
-    fn from(value: Value) -> Option<ReferenceValue> {
+impl From<Value> for VMResult<ReferenceValue> {
+    fn from(value: Value) -> VMResult<ReferenceValue> {
         match value.0 {
-            ValueImpl::Reference(reference) => Some(ReferenceValue::Reference(reference)),
-            ValueImpl::GlobalRef(reference) => Some(ReferenceValue::GlobalRef(reference)),
-            _ => None,
+            ValueImpl::Reference(reference) => Ok(ReferenceValue::Reference(reference)),
+            ValueImpl::GlobalRef(reference) => Ok(ReferenceValue::GlobalRef(reference)),
+            _ => {
+                let msg = format!("Cannot cast {:?} to ReferenceValue", value);
+                Err(VMStatus::new(StatusCode::INTERNAL_TYPE_ERROR).with_message(msg))
+            }
         }
     }
 }
 
-impl From<Value> for Option<Reference> {
-    fn from(value: Value) -> Option<Reference> {
+impl From<Value> for VMResult<Reference> {
+    fn from(value: Value) -> VMResult<Reference> {
         match value.0 {
-            ValueImpl::Reference(reference) => Some(reference),
-            _ => None,
+            ValueImpl::Reference(reference) => Ok(reference),
+            _ => {
+                let msg = format!("Cannot cast {:?} to Reference", value);
+                Err(VMStatus::new(StatusCode::INTERNAL_TYPE_ERROR).with_message(msg))
+            }
         }
     }
 }
 
-impl From<Value> for Option<GlobalRef> {
-    fn from(value: Value) -> Option<GlobalRef> {
+impl From<Value> for VMResult<GlobalRef> {
+    fn from(value: Value) -> VMResult<GlobalRef> {
         match value.0 {
-            ValueImpl::GlobalRef(reference) => Some(reference),
-            _ => None,
+            ValueImpl::GlobalRef(reference) => Ok(reference),
+            _ => {
+                let msg = format!("Cannot cast {:?} to GlobalReference", value);
+                Err(VMStatus::new(StatusCode::INTERNAL_TYPE_ERROR).with_message(msg))
+            }
```

Comment:
Honestly I'm not really a fan of these impls, they're too magical -- should they be expressed in a different way? This can be done in a different PR though.

Reply:
yeah a different PR for me, hope it's ok with you too.
I mentioned in the top level PR comment/description. The `From` impl is particularly odd.
I was suggesting a `Cast` private impl just to move to something more descriptive. And suggesting it would be private to the module given all external crate use and should use `Value::value_as<T>`. 
Not sure if you have suggestions here

Reply:
Yeah, I agree that a `Cast` trait that's kept as private or `pub(crate)` makes sense.